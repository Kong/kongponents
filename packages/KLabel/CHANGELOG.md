# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 6.1.9 (2021-11-18)

**Note:** Version bump only for package @kongponents/klabel





## 6.1.8 (2021-11-18)


### Bug Fixes

* **css:** box-shadow does not render on ios ([#420](https://github.com/Kong/kongponents/issues/420)) ([c0ba7cb](https://github.com/Kong/kongponents/commit/c0ba7cb90bf3c158a1baa6b879d33230359db830))





## 6.1.7 (2021-11-17)

**Note:** Version bump only for package @kongponents/klabel





## 6.1.6 (2021-11-17)


### Bug Fixes

* **ktextarea:** limit exceeded event ([#418](https://github.com/Kong/kongponents/issues/418)) ([f7cf421](https://github.com/Kong/kongponents/commit/f7cf421a4ded7f5b99eaa5a18d02a27d450020b7))





## 6.1.5 (2021-11-17)

**Note:** Version bump only for package @kongponents/klabel





## 6.1.4 (2021-11-17)


### Bug Fixes

* general fixes ([#415](https://github.com/Kong/kongponents/issues/415)) ([5b4bb48](https://github.com/Kong/kongponents/commit/5b4bb484d74e75779d39cf901b2cdaa23818e8e7))





## 6.1.3 (2021-11-17)

**Note:** Version bump only for package @kongponents/klabel





## [6.1.2](https://github.com/Kong/kongponents/compare/@kongponents/klabel@5.0.8...@kongponents/klabel@6.1.2) (2021-11-17)

Kongponents redesign - no changes to this component


## [5.0.8](https://github.com/Kong/kongponents/compare/v5.0.7...v5.0.8) (2021-11-15)

**Note:** Version bump only for package @kongponents/klabel





## [5.0.7](https://github.com/Kong/kongponents/compare/v5.0.6...v5.0.7) (2021-10-26)

**Note:** Version bump only for package @kongponents/klabel





## [5.0.6](https://github.com/Kong/kongponents/compare/v5.0.5...v5.0.6) (2021-10-26)

**Note:** Version bump only for package @kongponents/klabel





## [5.0.5](https://github.com/Kong/kongponents/compare/v5.0.4...v5.0.5) (2021-10-13)

**Note:** Version bump only for package @kongponents/klabel





## [5.0.4](https://github.com/Kong/kongponents/compare/v5.0.3...v5.0.4) (2021-10-06)

**Note:** Version bump only for package @kongponents/klabel





## [5.0.3](https://github.com/Kong/kongponents/compare/v5.0.2...v5.0.3) (2021-09-22)

**Note:** Version bump only for package @kongponents/klabel





## [5.0.2](https://github.com/Kong/kongponents/compare/v5.0.1...v5.0.2) (2021-09-15)

**Note:** Version bump only for package @kongponents/klabel





## [5.0.1](https://github.com/Kong/kongponents/compare/v5.0.0...v5.0.1) (2021-09-13)

**Note:** Version bump only for package @kongponents/klabel





# [5.0.0](https://github.com/Kong/kongponents/compare/v4.0.2...v5.0.0) (2021-09-13)

**Note:** Version bump only for package @kongponents/klabel





## [4.0.2](https://github.com/Kong/kongponents/compare/v4.0.1...v4.0.2) (2021-09-09)

**Note:** Version bump only for package @kongponents/klabel





## [4.0.1](https://github.com/Kong/kongponents/compare/v4.0.0...v4.0.1) (2021-09-07)

**Note:** Version bump only for package @kongponents/klabel





# 4.0.0 (2021-09-03)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))


### Features

* **forms:** Add KRadio & KCheckbox, add form css ([65158f2](https://github.com/Kong/kongponents/commit/65158f26ca6999a6fe47e0c58fe5a64aed8737d6))
* **klabel:** add for attribute for label ([#321](https://github.com/Kong/kongponents/issues/321)) ([05d2b4c](https://github.com/Kong/kongponents/commit/05d2b4c3fcafd42ee26e70d77e2f2e96dd441172))
* **klabel:** add help property with icon + tooltip ([#306](https://github.com/Kong/kongponents/issues/306)) ([d6b86d1](https://github.com/Kong/kongponents/commit/d6b86d109712a87da80434270c5cf1fe7ec19e1b))





# [0.5.0](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.4.0...@kongponents/klabel@0.5.0) (2021-08-24)


### Features

* **klabel:** add for attribute for label ([#321](https://github.com/Kong/kongponents/issues/321)) ([05d2b4c](https://github.com/Kong/kongponents/commit/05d2b4c3fcafd42ee26e70d77e2f2e96dd441172))





# [0.4.0](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.3.3...@kongponents/klabel@0.4.0) (2021-06-08)


### Features

* **klabel:** add help property with icon + tooltip ([#306](https://github.com/Kong/kongponents/issues/306)) ([d6b86d1](https://github.com/Kong/kongponents/commit/d6b86d109712a87da80434270c5cf1fe7ec19e1b))





## [0.3.3](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.3.2...@kongponents/klabel@0.3.3) (2020-03-06)

**Note:** Version bump only for package @kongponents/klabel





## [0.3.2](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.3.0...@kongponents/klabel@0.3.2) (2020-01-16)

**Note:** Version bump only for package @kongponents/klabel





# [0.3.0](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.2.4...@kongponents/klabel@0.3.0) (2019-11-14)


### Features

* **forms:** Add KRadio & KCheckbox, add form css ([65158f2](https://github.com/Kong/kongponents/commit/65158f26ca6999a6fe47e0c58fe5a64aed8737d6))





## [0.2.4](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.2.3...@kongponents/klabel@0.2.4) (2019-11-12)

**Note:** Version bump only for package @kongponents/klabel





## [0.2.3](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.2.2...@kongponents/klabel@0.2.3) (2019-11-11)

**Note:** Version bump only for package @kongponents/klabel





## [0.2.2](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.2.1...@kongponents/klabel@0.2.2) (2019-11-08)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))





## [0.2.1](https://github.com/Kong/kongponents/compare/@kongponents/klabel@0.2.0...@kongponents/klabel@0.2.1) (2019-11-08)

**Note:** Version bump only for package @kongponents/klabel
