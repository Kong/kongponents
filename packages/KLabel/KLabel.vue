<template functional>
  <label>
    <slot />
  </label>
</template>

<script>
export default {
  name: 'KLabel'
}
</script>

<style scoped>
label {
  display: inline-block;
  color: var(--tblack-85);
  font-weight: 500;
  font-family: var(--font-family-sans);
  font-size: var(--type-sm);
  line-height: var(--type-lg);
  margin-bottom: var(--spacing-xs);
}
</style>
