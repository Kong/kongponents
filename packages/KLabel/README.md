```js
<KLabel />
<KInput type="text" id="name" placeholder="Enter a Name" />
```
