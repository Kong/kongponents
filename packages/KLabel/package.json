{
  "name": "@kongponents/klabel",
  "version": "0.2.0",
  "description": "Form input label",
  "main": "KLabel.vue",
  "scripts": {
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme"
}
