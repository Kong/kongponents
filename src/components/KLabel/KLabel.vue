<template>
  <label class="k-input-label">
    <slot />
    <span
      v-if="required"
      class="is-required"
    >*</span>
    <KTooltip
      v-if="hasTooltip"
      v-bind="tooltipAttributes"
      class="label-tooltip"
      position-fixed
    >
      <KIcon
        :color="`var(--kui-color-text-neutral-weak, ${KUI_COLOR_TEXT_NEUTRAL_WEAK})`"
        hide-title
        icon="infoFilled"
        :size="KUI_ICON_SIZE_30"
      />
      <template #content>
        <slot name="tooltip">{{ info }}</slot>
      </template>
    </KTooltip>
  </label>
</template>

<script setup lang="ts">
import type { PropType } from 'vue'
import { computed, useSlots } from 'vue'
import KIcon from '@/components/KIcon/KIcon.vue'
import KTooltip from '@/components/KTooltip/KTooltip.vue'
import type { TooltipAttributes } from '@/types'
import { KUI_COLOR_TEXT_NEUTRAL_WEAK, KUI_ICON_SIZE_30 } from '@kong/design-tokens'

const props = defineProps({
  info: {
    type: String,
    default: '',
  },
  required: {
    type: Boolean,
    default: false,
  },
  tooltipAttributes: {
    type: Object as PropType<TooltipAttributes>,
    default: () => ({}),
  },
})

const slots = useSlots()

const hasTooltip = computed((): boolean => !!(props.info || slots.tooltip))
</script>

<style lang="scss" scoped>

.k-input-label {
  color: var(--kui-color-text, $kui-color-text);
  display: inline-flex;
  font-family: var(--kui-font-family-text, $kui-font-family-text);
  font-size: var(--kui-font-size-30, $kui-font-size-30);
  font-weight: var(--kui-font-weight-semibold, $kui-font-weight-semibold);
  line-height: var(--kui-line-height-30, $kui-line-height-30);
  margin-bottom: var(--kui-space-40, $kui-space-40);

  .is-required {
    color: var(--kui-color-text, $kui-color-text);
    font-size: var(--kui-font-size-30, $kui-font-size-30);
    font-weight: var(--kui-font-weight-semibold, $kui-font-weight-semibold);
    margin-left: var(--kui-space-20, $kui-space-20);
  }

  .label-tooltip {
    align-items: center;
    display: flex;

    :deep(.kong-icon) {
      &.kong-icon-info {
        cursor: pointer;
        height: 16px;
      }
    }

    :deep(.k-tooltip) {
      font-weight: var(--kui-font-weight-regular, $kui-font-weight-regular);

      code {
        background-color: var(--kui-color-background-neutral, $kui-color-background-neutral);
        color: var(--kui-color-text-inverse, $kui-color-text-inverse);
      }
    }
  }

  .kong-icon {
    margin-left: var(--kui-space-20, $kui-space-20);
  }
}
</style>
