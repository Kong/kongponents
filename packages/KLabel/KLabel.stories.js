import { storiesOf } from '../../storybook/node_modules/@storybook/vue'
import KLabel from './KLabel.vue'

storiesOf('KLabel', module)
  .add('Default', () => ({
    components: { KLabel },
    template: `<KLabel for="input_id">Label Text</KLabel>`
  }), {info: {}})
