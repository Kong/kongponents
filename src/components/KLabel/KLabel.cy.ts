import { mount } from 'cypress/vue'
import KLabel from '@/components/KLabel/KLabel.vue'

/**
 * ALL TESTS MUST USE testMode: true
 * We generate unique IDs for reference by aria properties. Test mode strips these out
 * allowing for successful snapshot verification.
 * props: {
 *   testMode: true
 * }
 */
describe('KLabel', () => {
  it('renders a plain label by default', () => {
    const text = 'Full Name'
    mount(KLabel, {
      props: {
        testMode: true,
      },
      slots: {
        default: () => text,
      },
    })

    cy.get('.k-input-label').should('have.text', text)
  })

  it('renders a tooltip when `help` is provided', () => {
    mount(KLabel, {
      props: {
        help: 'This is a tooltip',
        testMode: true,
      },
      slots: {
        default: () => 'Full Name',
      },
    })

    cy.get('.k-input-label .label-tooltip').should('not.be.empty')
  })

  it('renders an asterisk when `required` is true', () => {
    mount(KLabel, {
      props: {
        testMode: true,
        required: true,
      },
      slots: {
        default: () => 'Full Name',
      },
    })

    cy.get('.k-input-label .is-required').should('exist')
  })

  it('renders a tooltip when `info` is provided', () => {
    mount(KLabel, {
      props: {
        info: 'This is a tooltip',
        testMode: true,
      },
      slots: {
        default: () => 'Full Name',
      },
    })

    cy.get('.k-input-label .label-tooltip').should('not.be.empty')
  })

  it('passes the `for` attribute to label when `for` is provided', () => {
    const id = 'test-id'
    mount(KLabel, {
      props: {
        for: id,
        testMode: true,
      },
      slots: {
        default: () => 'Full Name',
      },
    })

    cy.get('.k-input-label').invoke('attr', 'for').should('eq', id)
  })
})
