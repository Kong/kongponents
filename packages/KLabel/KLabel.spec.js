import { mount } from '@vue/test-utils'
import KLabel from '@/KLabel/KLabel'

/**
 * ALL TESTS MUST USE testMode: true
 * We generate unique IDs for reference by aria properties. Test mode strips these out
 * allowing for successful snapshot verification.
 * propsData: {
 *   testMode: true
 * }
 */
describe('KLabel', () => {
  it('renders a plain label by default', () => {
    const wrapper = mount(KLabel, {
      propsData: {
        testMode: true
      },
      slots: {
        default: 'Full Name'
      }
    })

    expect(wrapper.find('label').text()).toBe('Full Name')
  })

  it('renders a tooltip when `help` is provided', () => {
    const wrapper = mount(KLabel, {
      propsData: {
        help: 'This is a tooltip',
        testMode: true
      },
      slots: {
        default: 'Full Name'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a tooltip when `info` is provided', () => {
    const wrapper = mount(KLabel, {
      propsData: {
        info: 'This is a tooltip',
        testMode: true
      },
      slots: {
        default: 'Full Name'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('passes the `for` attribute to label when `for` is provided', () => {
    const wrapper = mount(KLabel, {
      propsData: {
        for: 'test-id',
        testMode: true
      },
      slots: {
        default: 'Full Name'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
