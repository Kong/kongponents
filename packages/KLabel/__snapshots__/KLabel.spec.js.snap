// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KLabel passes the \`for\` attribute to label when \`for\` is provided 1`] = `<label for="test-id" class="k-input-label">Full Name</label>`;

exports[`KLabel renders a tooltip when \`help\` is provided 1`] = `
<label class="k-input-label">
  <div id="dcd15dd1-22e0-11ec-bf28-e7912c7d1fab" aria-controls="dcd15dd0-22e0-11ec-bf28-e7912c7d1fab" role="button" tab-index="0" class="label-tooltip">Full Name <svg height="16" width="16" viewBox="0 0 24 24" role="img" class="kong-icon kong-icon-help">
      <!---->
      <g>
        <!---->
        <!---->
        <path d="M12 2a10.01 10.01 0 0 1 0 20 10 10 0 0 1 0-20Zm-.18 12h.79c.23 0 .35-.1.4-.32.05-.21.1-.42.18-.62.15-.32.43-.5.72-.68 1.18-.74 1.87-1.75 1.7-3.2-.12-.91-.55-1.61-1.32-2.1a4.98 4.98 0 0 0-3.91-.3c-1.07.4-1.72 1.13-1.8 2.31-.01.21.1.37.32.38.52.01 1.04.03 1.57.03.25 0 .42-.14.5-.38.14-.51.51-.72 1.01-.77.57-.04 1.08.3 1.2.8.2.75.03 1.42-.56 1.94-.35.31-.73.58-1.1.87-.49.4-.83.9-.87 1.56-.02.31.1.48.42.48h.75Zm-.02 1.56h-.67c-.3 0-.52.22-.52.53v.8c0 .35.22.57.56.57h1.27c.32 0 .53-.2.55-.53a8.8 8.8 0 0 0 0-.85c-.02-.33-.23-.52-.57-.53h-.62Z" fill="#A3BBCC"></path>
      </g>
    </svg>
    <div id="dcd15dd0-22e0-11ec-bf28-e7912c7d1fab" role="region" tab-index="0" class="k-popover kooltip mt-2  hide-caret" style="width: auto; display: none;" name="fade">
      <!---->
      <div class="k-popover-content">
        <div role="tooltip">This is a tooltip</div>
      </div>
      <!---->
    </div>
  </div>
</label>
`;

exports[`KLabel renders a tooltip when \`info\` is provided 1`] = `
<label class="k-input-label">
  <div id="dcd52e61-22e0-11ec-bf28-e7912c7d1fab" aria-controls="dcd52e60-22e0-11ec-bf28-e7912c7d1fab" role="button" tab-index="0" class="label-tooltip">Full Name <svg height="16" width="16" viewBox="0 0 16 16" role="img" class="kong-icon kong-icon-info">
      <!---->
      <g>
        <!---->
        <!---->
        <path fill="#A3BBCC" fill-rule="evenodd" d="M16 8c0-4.418278-3.58172-8-8-8S0 3.581722 0 8s3.58172 8 8 8 8-3.581722 8-8zm-1.5 0c0 3.5898509-2.91015 6.5-6.5 6.5S1.5 11.5898509 1.5 8 4.41015 1.5 8 1.5s6.5 2.9101491 6.5 6.5zM6 12h4v-1H9V7H6v1h1v3H6v1zm1-6.5046844C7 5.7740451 7.21404 6 7.50468 6h.99064C8.77405 6 9 5.785965 9 5.4953156v-.9906312C9 4.2259549 8.78596 4 8.49532 4h-.99064C7.22595 4 7 4.214035 7 4.5046844v.9906312z"></path>
      </g>
    </svg>
    <div id="dcd52e60-22e0-11ec-bf28-e7912c7d1fab" role="region" tab-index="0" class="k-popover kooltip mt-2  hide-caret" style="width: auto; display: none;" name="fade">
      <!---->
      <div class="k-popover-content">
        <div role="tooltip">This is a tooltip</div>
      </div>
      <!---->
    </div>
  </div>
</label>
`;
